Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -198,6 +221,264 @@ void WebExtension::populateContentSecurityPolicyStringsIfNeeded()
+        return nullString();
+
+    if (m_backgroundScriptPaths.size())
+        return nullString();
+
+    bool isServiceWorker = backgroundContentIsServiceWorker();
+    bool usesModules = backgroundContentUsesModules();
+
+    Vector<String> scripts;
+    for (auto& scriptPath : m_backgroundScriptPaths) {
+        StringBuilder format;
+        if (isServiceWorker) {
+            if (usesModules) {
+                format.append("import \""_s, scriptPath, "\";"_s);
+
+                scripts.append(format.toString());
+                continue;
+            }
+
+            format.append("importScripts(\""_s, scriptPath, "\");"_s);
+
+            scripts.append(format.toString());
+            continue;
+        }
+
+        format.append("<script"_s);
+        if (usesModules)
+            format.append(" type=\"module\" "_s);
+        format.append("src=\""_s, scriptPath, "\"></script>"_s);
+
+        scripts.append(format.toString());
+    }
+
+    StringBuilder generatedBackgroundContent;
+    if (!isServiceWorker)
+        generatedBackgroundContent.append("<!DOCTYPE html>\n<body>\n"_s);
+
+    for (auto scriptPath : scripts)
+        generatedBackgroundContent.append(scriptPath, "\n"_s);
+
+    if (!isServiceWorker)
+        generatedBackgroundContent.append("\n</body>"_s);
+
+    m_generatedBackgroundContent = generatedBackgroundContent.toString();
+    return m_generatedBackgroundContent;
+}
+
+void WebExtension::populateBackgroundPropertiesIfNeeded()
+{
+    if (m_parsedManifestBackgroundProperties)
+        return;
+
+    m_parsedManifestBackgroundProperties = true;
+
+    RefPtr manifestObject = this->manifestObject();
+    if (!manifestObject)
+        return;
+
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/manifest.json/background
+
+    RefPtr backgroundManifestObject = manifestObject->getObject(backgroundManifestKey);
+    if (!backgroundManifestObject || !backgroundManifestObject->size()) {
+        if (manifestObject->getValue(backgroundManifestKey))
+            recordError(createError(Error::InvalidBackgroundContent));
+        return;
+    }
+
+    m_backgroundPagePath = backgroundManifestObject->getString(backgroundPageManifestKey);
+    m_backgroundServiceWorkerPath = backgroundManifestObject->getString(backgroundServiceWorkerManifestKey);
+    m_backgroundContentUsesModules = (backgroundManifestObject->getString(backgroundPageTypeKey) == backgroundPageTypeModuleValue);
+
+    if (RefPtr backgroundScriptPaths = backgroundManifestObject->getArray(backgroundScriptsManifestKey)) {
+        backgroundScriptPaths = filterObjects(*backgroundScriptPaths, [](auto& value) {
+            return !value.asString().isEmpty();
+        });
+
+        m_backgroundScriptPaths = makeStringVector(*backgroundScriptPaths);
+    }
+
+    Vector<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
```

Comment:
This could have stayed as `HashSet` since all we do with it is call `contains()`.

Reply:
```suggestion
    HashSet<String> supportedEnvironments = { backgroundDocumentManifestKey, backgroundServiceWorkerManifestKey };
```